pynitrokey/nk3/otp_app.py

```diff
@@ -243,3 +265,101 @@ def verify_code(self, cred_id: bytes, code: int) -> bool:
         ]
         res = self._send_receive(Instruction.VerifyCode, structure=structure)
         return res.hex() == "7700"
+
+    def set_code_cli(self, passphrase: str) -> None:
+        """
+        Set the code with the defaults as suggested in the protocol specification:
+        - https://developers.yubico.com/OATH/YKOATH_Protocol.html
+        """
+        secret = self.get_secret_for_passphrase(passphrase)
+        challenge = token_bytes(8)
+        response = self.get_response_for_secret(challenge, secret)
+        self.set_code(secret, challenge, response)
+
+    def get_secret_for_passphrase(self, passphrase: str) -> bytes:
+        #   secret = PBKDF2(USER_PASSPHRASE, DEVICEID, 1000)[:16]
+        # salt = self.select().name
+        # FIXME use the proper SALT
+        # FIXME USB/IP Sim changes its ID after each reset and after setting the code (??)
+        salt = b"a" * 8
+        secret = pbkdf2_hmac("sha256", passphrase.encode(), salt, 1000)
+        return secret[:16]
+
+    def get_response_for_secret(self, challenge: bytes, secret: bytes) -> bytes:
+        response = hmac.HMAC(key=secret, msg=challenge, digestmod="sha1").digest()
+        return response
+
+    def set_code(self, key: bytes, challenge: bytes, response: bytes) -> None:
+        """
+        Set or clear the passphrase used to authenticate to other commands. Raw interface.
+        :param key: User passphrase processed through PBKDF2(ID,1000), and limited to the first 16 bytes.
+        :param challenge: The current challenge taken from the SELECT command.
+        :param response: The data calculated on the client, as a proof of a correct setup.
+        """
+        algo = Algorithm.Sha1.value
+        kind = Kind.Totp.value
+        structure = [
+            tlv8.Entry(Tag.Key.value, bytes([kind | algo]) + key),
+            tlv8.Entry(Tag.Challenge.value, challenge),
+            tlv8.Entry(Tag.Response.value, response),
+        ]
+        self._send_receive(Instruction.SetCode, structure=structure)
+
+    def clear_code(self) -> None:
+        """
+        Clear the passphrase used to authenticate to other commands.
+        """
+        structure = [
+            tlv8.Entry(Tag.Key.value, bytes()),
+        ]
+        self._send_receive(Instruction.SetCode, structure=structure)
+
+    def validate_cli(self, passphrase: str) -> None:
+        """
+        Authenticate using a passphrase
+        """
+        stat = self.select()
+        assert stat.algorithm == bytes([Algorithm.Sha1.value])
+        challenge = stat.challenge
+        if challenge is None:
+            raise RuntimeWarning("Challenge is not available")
```

Comment:
This is a user-facing error message, right?  In this case, something like “No passphrase set” might be more helpful.

Reply:
Good spot. It was not intended initially, but it's worth to have it properly described anyway. Elaborated and refactored a bit.